jr-objects/src/main/java/com/fasterxml/jackson/jr/ob/impl/ValueReaderLocator.java

```diff
@@ -452,7 +452,18 @@ protected BeanReader _resolveBeanForDeser(Class<?> raw, POJODefinition beanDef)
         final Map<String, BeanPropertyReader> propMap;
         Map<String, String> aliasMapping = null;
 
-        boolean isRecord = RecordsHelpers.isRecordType(raw);
+        final Map<String, Integer> ctorParameterIndexMap = new HashMap<>();
```

Comment:
Ok. While this works, technically, I think I'd like original properties to be returned in declaration order, so that index would match. I am not (yet) sure why the order gets shuffled, will try to read code further.

Reply:
Oh, something probably sorts properties alphabetically which leads to discrepancy.
So that serialization order is alphabetic by default.

There's also then the question wrt Records: should they

1. Follow POJO logic of sorting by alphabetic order, OR
2. Use Record declaration order

To me (2) is probably more often default to use, but this probably should be configurable option.
And for backwards compatibility should probably default to (1) and allow (2) via configuration.

Reply:
Ok it all happens in `BeanPropertyIntrospector`... due to

    Map<String,PropBuilder> propsByName = new TreeMap<>();

changing of which is not trivial I suppose. And problem being that Record support is based on Fields and/or Methods and not just the canonical constructor.
But I think it can be hacked a bit to use declaration order if

1. Type is record AND
2. We are introspection for Deserialization

(and with new possible feature, also for Serialization -- but that'd be for 2.19)

Reply:
The record constructor does return them in declaration order, the 
```
final List<POJODefinition.Prop> rawProps = beanDef.getProperties();
``` 
(which is probably created through the `BeanPropertyIntrospector`) returns them in alphabetic order.

Changing how that internally works seems to me like something that should be proposed in a separate PR as it might impact more than just this deserialisation part and might require some rethinking of (de)serialisation of records?

Reply:
@Giovds Right -- what I may do is to take tests from here, but make code change differently.
I realized that the way Record introspection works is bit different than I'd like -- trying to be minimal wrt changes from POJOs -- but I think I can get ordering to work in a way that no sorting should be needed.